SimulationAgent/Simulation/DeviceStatusLogic/UpdateReportedProperties.cs

```diff
@@ -0,0 +1,172 @@
+using System.Threading;
+
+namespace Microsoft.Azure.IoTSolutions.DeviceSimulation.SimulationAgent.Simulation.DeviceStatusLogic
+{
+
+    public class UpdateReportedProperties : IDeviceStatusLogic
+    {
+        // When connecting to IoT Hub, timeout after 10 seconds
+        private static readonly TimeSpan connectionTimeout = TimeSpan.FromSeconds(10);
+
+        private IDevices devices;
+        private string deviceId;
+        private DeviceModel deviceModel;
+
+        private readonly ILogger log;
+
+        // Ensure that setup is called once and only once (which helps also detecting thread safety issues)
+        private bool setupDone = false;
+
+        public UpdateReportedProperties(
+            IDevices devices,
+            ILogger logger)
+        {
+            this.log = logger;
+            this.devices = devices;
+        }
+
+        public void Run(object context)
+        {
+            try { 
+
+                this.ValidateSetup();
+
+                var actor = (IDeviceActor)context;
+                if (actor.CancellationToken.IsCancellationRequested)
+                {
+                    actor.Stop();
+                    return;
+                }
+
+                //TODO: Here we should pause the timer in case the device takes too long to pull from the hub
+
+                this.log.Debug(
+                    "Checking for desired property updates & updated reported properties",
+                    () => new { this.deviceId, deviceState = actor.DeviceState
+                    });
+
+                // Get device
+                var device = this.GetDevice(actor.CancellationToken);
+                lock (actor.DeviceState)
+                {
+                    // check for differences between reported/desired properties,
+                    // update reported properties with any state changes (either from desired prop changes, methods, etc.)
+                    if (this.ChangeTwinPropertiesToMatchDesired(device, actor.DeviceState)
+                        || this.ChangeTwinPropertiesToMatchActorState(device, actor.DeviceState))
+                        actor.BootstrapClient.UpdateTwinAsync(device).Wait((int)connectionTimeout.TotalMilliseconds);
+                }
+
+
+                // Move state machine forward to start sending telemetry messages if needed
+                if (actor.ActorStatus == Status.UpdatingReportedProperties)
+                {
+                    actor.MoveNext();
+                }
+                else
+                {
+                    this.log.Debug(
+                        "Already moved state machine forward, continuing to check for desired property changes",
+                        () => new { this.deviceId });
+                }
+
+            }
+            catch (Exception e)
+            {
+                this.log.Error("UpdateReportedProperties failed",
+                    () => new { this.deviceId, e.Message, Error = e.GetType().FullName});
```

Comment:
just log "e", the logger takes care of logging the interesting stuff

Reply:
done